### PR TITLE
feat: add dividends list page

### LINKED ISSUE #769: Product: View all Dividends
**Goal:**
As a products admin, I want to view all the dividends for a particular share product,
[https://demo.mifos.io/#/dividends/82/](https://demo.mifos.io/#/dividends/82/)

Screenshot from 2020-05-15 00-49-59

### PR DESCRIPTION
**Pull request by stableapple**
## Description
Added view for dividends in share product.

## Related issues and discussion
Fixes #769

## Screenshots, if any
Screenshot 2020-05-31 11:01:41

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

**Comment by stableapple** [comment]
@karantakalkar Please review.

File: src/app/products/products-routing.module.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ import { ManageTaxGroupsComponent } from './manage-tax-groups/manage-tax-groups.
 import { ViewTaxComponentComponent } from './manage-tax-components/view-tax-component/view-tax-component.component';
 import { CreateTaxComponentComponent } from './manage-tax-components/create-tax-component/create-tax-component.component';
 import { EditTaxComponentComponent } from './manage-tax-components/edit-tax-component/edit-tax-component.component';
-
+import { DividendsComponent } from './share-products/dividends-share-product/dividends.components';
```
Comment (karantakalkar) [inline src/app/products/products-routing.module.ts:42]: Leave a line gap b/w components and resolvers.

File: src/app/products/products-routing.module.ts [hunk 2]
```diff
@@ -116,8 +116,8 @@ const routes: Routes = [
                   data: { title: extract('Edit Loan Product'), breadcrumb: 'Edit', routeParamBreadcrumb: false },
                   resolve: {
                     loanProductAndTemplate: LoanProductAndTemplateResolver
-                  }
-                }
+                  },
+                },
```
Comment (karantakalkar) [inline src/app/products/products-routing.module.ts:120]:
```suggestion
                  }
                }
```

File: src/app/products/products-routing.module.ts [hunk 3]
```diff
@@ -200,7 +200,12 @@ const routes: Routes = [
                   data: { title: extract('Edit Share Product'), breadcrumb: 'Edit', routeParamBreadcrumb: false },
                   resolve: {
                     shareProductAndTemplate: ShareProductAndTemplateResolver
-                  }
+                  },
```
Comment (karantakalkar) [inline src/app/products/products-routing.module.ts:203]: Add comma only if there is another object below it.

File: src/app/products/products-routing.module.ts [hunk 4]
```diff
@@ -200,7 +200,12 @@ const routes: Routes = [
...
+                },
+                {
+                  path: 'dividends',
+                  data: { title: extract('Dividends Share Product'), breadcrumb: 'id'},
```
Comment (karantakalkar) [inline src/app/products/products-routing.module.ts:207]:
```suggestion
                  data: { title: extract('Share Product Dividends'), breadcrumb: 'Dividends', routeParamBreadcrumb: false },
```

File: src/app/products/products-routing.module.ts [hunk 5]
```diff
@@ -432,6 +437,6 @@ const routes: Routes = [
     FloatingRatesResolver,
     TaxComponentTemplateResolver,
     EditTaxComponentComponent
-  ]
+  ],
```
Comment (karantakalkar) [inline src/app/products/products-routing.module.ts:440]:
```suggestion
  ]
```

File: src/app/products/products.module.ts [hunk 6]
```diff
@@ -74,6 +74,7 @@ import { EditChargeComponent } from './charges/edit-charge/edit-charge.component
 import { MatFormFieldModule, MatSelectModule, MatOptionModule } from '@angular/material';
 import { FormsModule, ReactiveFormsModule } from '@angular/forms';
 
+import { DividendsComponent } from './share-products/dividends-share-product/dividends.components';
```
Comment (karantakalkar) [inline src/app/products/products.module.ts:77]: Rename this component as `ShareProductsDividendsComponent`, import it where other ShareProduct imports are.

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 7]
```diff
@@ -0,0 +1,52 @@
+<div class="container m-b-20" fxLayout="row" fxLayoutAlign="end" fxLayoutGap="20px">
+    <button mat-raised-button color="primary">
+      <fa-icon icon="plus"></fa-icon>&nbsp;&nbsp;
+      Create Dividend
+    </button>
+</div>
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:6]:
```suggestion
</div>

```

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    <mat-paginator [pageSizeOptions]="[10, 25, 50, 100]" showFirstLastButtons></mat-paginator>
+
+  </div>
+
+</div>
+
+
+
+
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:52]: Only 1 new line at EOF

File: src/app/products/share-products/dividends-share-product/dividends.component.scss [hunk 9]
```diff
@@ -0,0 +1,7 @@
+table {
+    width: 100%;
+
+    .select-row:hover {
+      cursor: pointer;
+    }
+  }
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.scss:7]: Add a new line at EOF.

File: src/app/products/share-products/dividends-share-product/dividends.components.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+  @ViewChild(MatSort) sort: MatSort;
+
+  constructor() { }
+
+
+  ngOnInit() {
+    this.dataSource = new MatTableDataSource(this.dividendsData);
+    this.dataSource.paginator = this.paginator;
+    this.dataSource.sort = this.sort;
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.components.ts:24]: Shift this to a new function set dividends

File: src/app/products/share-products/dividends-share-product/dividends.components.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
+import { Component, OnInit, ViewChild } from '@angular/core';
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.components.ts:1]: Add jsdoc comments throughout the file, take reference from accounting section on how commenting is done.

File: src/app/products/share-products/dividends-share-product/dividends.components.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+})
+export class DividendsComponent implements OnInit {
+  dividendsData: any;
+  displayedColumns: string[] = ['name', 'startDividend', 'endDividend', 'status'];
+  dataSource: MatTableDataSource<any>;
+
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+  @ViewChild(MatSort) sort: MatSort;
+
+  constructor() { }
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.components.ts:18]: You haven't connected the component to the API?, you have to construct API call in the service, use a resolver for it and then subscribe to data from route, see other simple table implementations in the app.
For which api endpoint to use you can open network tab in browser navigate to the same page in demo.mifos.io or staging.mios.io and see which API request got triggered upon navigation.

**Comment by stableapple** [comment]
@karantakalkar Made the changes. Please review.

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+      </ng-container>
+
+      <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
+      <tr mat-row *matRowDef="let row; columns: displayedColumns;" [routerLink]="[row.id]" class="select-row"></tr>
+
+    </table>
+
+    <mat-paginator [pageSizeOptions]="[10, 25, 50, 100]" showFirstLastButtons></mat-paginator>
+
+</div>
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:50]:
```suggestion
  </div>
```

File: src/app/products/share-products/dividends-share-product/dividends.component.scss (hunk 9 above)
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.scss:7]:
```suggestion
 }
```

File: src/app/products/products.service.ts [hunk 14]
```diff
@@ -90,6 +90,9 @@ export class ProductsService {
     return this.http.put(`/products/share/${shareProductId}`, shareProduct);
   }
 
+  getDividends(shareProductId: string): Observable<any> {
+    return this.http.get(`/shareproducts/${shareProductId}/dividends`);
```
Comment (karantakalkar) [inline src/app/products/products.service.ts:94]:
```suggestion
    return this.http.get(`/shareproduct/${shareProductId}/dividend`);
```
Please check the endpoint, navigation to component gives 404 in API response

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+      <ng-container matColumnDef="startDividend">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Dividend Period Start Date </th>
+        <td mat-cell *matCellDef="let dividends">{{ dividends.startDividend }}</td>
+      </ng-container>
+
+      <ng-container matColumnDef="endDividend">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Dividend Period End Date </th>
+        <td mat-cell *matCellDef="let dividends">{{ dividends.endDividend }}</td>
+      </ng-container>
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:33]:
```suggestion
      </ng-container>
      
```

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+      <ng-container matColumnDef="amount">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Amount </th>
+        <td mat-cell *matCellDef="let dividends"> {{ dividends.amount }}</td>
+      </ng-container>
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:37]:
```suggestion
      </ng-container>
      
```

File: src/app/products/share-products/dividends-share-product/dividends.components.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+})
+export class ShareProductsDividendsComponent implements OnInit {
+  dividendsData: any;
+  displayedColumns: string[] = ['name', 'startDividend', 'endDividend', 'status'];
+  dataSource: MatTableDataSource<any>;
+
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+  @ViewChild(MatSort) sort: MatSort;
+
+  constructor(private route: ActivatedRoute) {
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.components.ts:19]: Please comment every function and variable just as in reference PR I mentioned before.

File: src/app/products/share-products/dividends-share-product/dividends.components.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+    this.route.data.subscribe((data: {dividends: any}) => {
+      this.dividendsData = data.dividends.pageItems;
+    });
+   }
+
+
+  ngOnInit() {
+    this.setDividends();
+
+  }
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.components.ts:29]:
```suggestion
    this.setDividends();
  }
  
```
Indentations are very important, many people will see your code when it's merged.

File: src/app/products/share-products/dividends-share-product/dividends.components.ts [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    this.dataSource = new MatTableDataSource(this.dividendsData);
+    this.dataSource.paginator = this.paginator;
+    this.dataSource.sort = this.sort;
+  }
+
+  applyFilter(filterValue: string) {
+    this.dataSource.filter = filterValue.trim().toLowerCase();
+  }
+}
+
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.components.ts:40]: this is 2 line at EOF, 1 line at EOF looks like scss file of this component

File: src/app/products/share-products/dividends-share-product/dividends.resolver.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+import { ProductsService } from '../../products.service';
+
+/**
+ * Share products data resolver.
+ */
+@Injectable()
+export class DividendsResolver implements Resolve<Object> {
+
+  /**
+   *
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.resolver.ts:18]:
```suggestion
```
Take care of redundant lines, commas, spaces etc.

File: src/app/products/share-products/dividends-share-product/dividends.resolver.ts [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+   * Returns the share products data.
+   * @returns {Observable<any>}
+   */
+  resolve(route: ActivatedRouteSnapshot): Observable<any> {
+    const id = route.paramMap.get('id');
+    return this.productsService.getDividends(id);
+  }
+
+}
+
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.resolver.ts:33]: Only 1 line at EOF this is 2.

**Comment by stableapple** [comment]
@karantakalkar Made the suggested changes. Please review

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+  <div fxLayout="row">
+    <mat-form-field fxFlex>
+      <mat-label>Filter</mat-label>
+      <input matInput (keyup)="applyFilter($event.target.value)">
+    </mat-form-field>
+  </div>
+
+  <div class="mat-elevation-z8">
+
```
[the remark this answers is not in the document]
Comment (karantakalkar) [reply src/app/products/share-products/dividends-share-product/dividends.component.html:17]: Make corrections for every column.

**Comment by karantakalkar** [comment]
Change commit message to `feat: add dividends list page` as well.

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+      </ng-container>
+
+      <ng-container matColumnDef="amount">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Amount </th>
+        <td mat-cell *matCellDef="let dividend"> {{ dividend.amount }}</td>
+      </ng-container>
+
+      <ng-container matColumnDef="status">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Status </th>
+        <td mat-cell *matCellDef="let dividend">{{ dividend.status }}</td>
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:42]:
```suggestion
        <td mat-cell *matCellDef="let dividend">{{ dividend.status.value }}</td>
```
I guess you couldn't find a working test case? change url to `demo.mifos.io` share product `8` has a list of dividends.

Comment (stableapple) [reply src/app/products/share-products/dividends-share-product/dividends.component.html:42]: It shows invalid user details when I login.

Comment (karantakalkar) [reply src/app/products/share-products/dividends-share-product/dividends.component.html:42]: Yes, there were some changes made to auth recently, login using dev.mifos.io check the remember me checkbox, once you are logged in change it to demo.mifos.io.

**Comment by stableapple** [comment]
@karantakalkar Updated the screenshot. Please review.

File: src/app/products/share-products/dividends-share-product/dividends.components.ts [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+    this.dataSource.sort = this.sort;
+  }
+
+  /**
+   * Filters data in dividends table based on passed value.
+   * @param {string} filterValue Value to filter data.
+   */
+  applyFilter(filterValue: string) {
+    this.dataSource.filter = filterValue.trim().toLowerCase();
+  }
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.components.ts:60]:
```suggestion
  }
 
```

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 25]
```diff
@@ -0,0 +1,54 @@
+<div class="container m-b-20" fxLayout="row" fxLayoutAlign="end" fxLayoutGap="20px">
+    <button mat-raised-button color="primary">
+      <fa-icon icon="plus"></fa-icon>&nbsp;&nbsp;
+      Create Dividend
+    </button>
+</div>
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:6]:
```suggestion
</div>

```

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+    <table mat-table [dataSource]="dataSource" matSort>
+
+      <ng-container matColumnDef="name">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Name </th>
+        <td mat-cell *matCellDef="let dividends">{{ dividends.productData.name }}</td>
+      </ng-container>
+
+      <ng-container matColumnDef="dividendPeriodStartDate">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Dividend Period Start Date </th>
+        <td mat-cell *matCellDef="let dividends">{{ dividends.dividendPeriodStartDate | date }}</td>
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:27]:
```suggestion
        <td mat-cell *matCellDef="let dividend">{{ dividend.dividendPeriodStartDate | date }}</td>
```
please make this change for all mat cell defs

File: src/app/products/share-products/dividends-share-product/dividends.component.html [hunk 27]
```diff
@@ -0,0 +1,54 @@
+<div class="container m-b-20" fxLayout="row" fxLayoutAlign="end" fxLayoutGap="20px">
+    <button mat-raised-button color="primary">
+      <fa-icon icon="plus"></fa-icon>&nbsp;&nbsp;
+      Create Dividend
```
Comment (karantakalkar) [inline src/app/products/share-products/dividends-share-product/dividends.component.html:4]:
```suggestion
      Initiate Dividend
```

**Comment by stableapple** [comment]
@karantakalkar Build error due to timeout. Kindly review the changes.

**Comment by karantakalkar** [comment]
Screenshot from 2020-05-31 17-12-52
This is what I meant by adding Fixes.

**Comment by stableapple** [comment]
@karantakalkar Changed the commit message.

**Comment by karantakalkar** [comment]
Screenshot from 2020-05-31 17-46-21
Use git commit --amend, type in `feat: add dividends list page` leave a line and type `Fixes: #769`

**Comment by stableapple** [comment]
@karantakalkar Thanks for your help :)

**Pull request merged**